zenfs.mk

```diff
@@ -1,3 +1,27 @@
 zenfs_SOURCES = fs/fs_zenfs.cc fs/zbd_zenfs.cc fs/io_zenfs.cc
-zenfs_HEADERS = fs/fs_zenfs.h fs/zbd_zenfs.h fs/io_zenfs.h
+zenfs_HEADERS = fs/fs_zenfs.h fs/zbd_zenfs.h fs/io_zenfs.h fs/version.h
 zenfs_LDFLAGS = -lzbd -u zenfs_filesystem_reg
+
+ZENFS_ROOT_DIR := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
+
+ZENFS_GIT_HEAD := $(realpath $(shell if [ -d "$(ZENFS_ROOT_DIR)/.git" ]; then \
+echo "$(ZENFS_ROOT_DIR)/.git/HEAD"; \
+else echo "$(ZENFS_ROOT_DIR)/$$( cat $(ZENFS_ROOT_DIR)/.git | awk -F ' ' '{print $$2}' )/HEAD"; \
+fi))
+
+# 'git describe --dirty' will output a version in that looks like "v0.1.0-12-g3456789-dirty".
+# The individual version segements are parsed and written into the version.h file when the HEAD has changed.
+# Last line: Do not track changes of version.h and avoid problem where changes to version.h block switching 
+# the branch https://stackoverflow.com/questions/1248029/git-pull-error-entry-foo-not-uptodate-cannot-merge
+plugin/zenfs/fs/version.h: $(ZENFS_GIT_HEAD)
+	$(info Generating ZenFS version)
+	$(eval ZENFS_VERSION := $(shell cd $(ZENFS_ROOT_DIR) && git describe --dirty))
+	$(eval ZENFS_VERSION_MAJOR := $(shell echo $(ZENFS_VERSION) | awk -F . '{ print substr($$1,2); }'))
+	$(eval ZENFS_VERSION_MINOR := $(shell echo $(ZENFS_VERSION) | awk -F . '{ print $$2 }'))
+	$(eval ZENFS_PATCH_AND_SUFFIX := $(shell echo $(ZENFS_VERSION) | awk -F . '{ print $$3 }'))
+	$(eval ZENFS_VERSION_PATCH := $(shell echo $(ZENFS_PATCH_AND_SUFFIX) | awk -F - '{ print $$1 }'))
+	$(eval ZENFS_VERSION_COMMIT := $(shell echo $(ZENFS_PATCH_AND_SUFFIX) | awk -F - '{ print substr($$3,2); }'))
+	$(eval ZENFS_VERSION_DIRTY := $(shell echo $(ZENFS_PATCH_AND_SUFFIX) | awk -F - 'NF == 4 { print "!" }'))
+	$(eval ZENFS_VERSION_COMMIT := $(ZENFS_VERSION_COMMIT)$(ZENFS_VERSION_DIRTY))
+	printf '#pragma once\n#define ZENFS_VERSION "%s"\n#define ZENFS_VERSION_MAJOR "%s"\n#define ZENFS_VERSION_MINOR "%s"\n#define ZENFS_VERSION_PATCH "%s"\n#define ZENFS_VERSION_COMMIT "%s"' $(ZENFS_VERSION) $(ZENFS_VERSION_MAJOR) $(ZENFS_VERSION_MINOR) $(ZENFS_VERSION_PATCH) $(ZENFS_VERSION_COMMIT) > $(ZENFS_ROOT_DIR)/fs/version.h
+	$(shell cd $(ZENFS_ROOT_DIR) && git update-index --assume-unchanged $(ZENFS_ROOT_DIR)/fs/version.h && git update-index --really-refresh > /dev/null)
```

Comment:
If possible, I think this shell kung fu should go into a separate shell script. It clutters the Makefile and makes it difficult to read.

Reply:
Separated the parsing into `generate-version.sh` and improved dependencies for building `version.h`